text/0069-api-v2.md

```diff
@@ -0,0 +1,172 @@
+# TiKV API V2
+
+## Motivation
+
+`API V2` is a set of breaking changes that aim to solve serval issues with current RawKV (hereafter referred to as `API V1`):
+
+1. RawKV is not safe to use along with TxnKV. By solving this, TiDB will be able to support RawKV as the table's storage engine, which will enrich TiDB's use cases.
+2. RawKV TTL is controlled by Store configuration. Switching the configuration will cause data corruption in silence.
+3. RawKV TTL is encoded into the value by appending 8-bytes UNIX timestamp to the end of the value, therefore it's hard to introduce other encode afterward.
+4. RawKV compare_and_swap is easy to make mistake.
+5. It could be nice if we can deploy multiple applications on one TiKV cluster.
+
+## Detailed Design
+
+#### New key-value codec
+
+This RFC introduces a new key encode to RawKV and TxnKV, and a new value encode to RawKV, which will allow the RawKV to be used along with TxnKV and also allow TiKV to flexibly add meta, e.g. the TTL, to a RawKV value.
+
+In addition, keys will be contained in keyspaces, where the keys in different keyspace are totally independant. If keyspace is not specified, the keyspace 'default' will be used.
+
+The `API V2` is enabled by a switch on PD. Since it changes the storage encode, it will be not compatible to switch between `API V1` and `API V2` while there are non-TiDB data in TiKV. TiDB data is specially treated in order to not be affected by the change.
+
+#### Key Encode
+
+Once the `API V2` is enabled, the key will be starting with either:
+
+1. `m` or `t`: TxnKV key. Used by TiDB.
+2. `x{keyspace prefix id}`: TxnKV key.
+3. `r{keyspace prefix id}`: RawKV key.
+4. `a{keyspace prefix id}`: RawKV atomic key, on which you can use `CompareAndSwap`.
```

Comment:
> is it possible to add mvcc support for rawkv key-values, which can help us simplify the incremental data replication like TiCDC?

This is a big topic. Ignoring the bunch of reasons, I don't think mvcc without transaction control has any useful scenario.

> why not setting the {keyspace prefix id} as the common prefix for transactional and non-transactional key-values like the following:

Previously rawkv and txnkv use different encoding (MCE), and to avoid a region from spanning over two modes, PD should always split the keys between each modes. But now the rawkv uses MCE too in API V2, so there is no any reason to put mode prefix before the kepsapce id.

> achieve better performance when accessing both transactional and non-transaction key-values

This is not practically true. because in each request we can only access one region by one mode, we can not benefit from putting data in different modes into the same region.

>physically isolate the storage and computation on different key-spaces via placement rule

Reasonable.

Reply:
how about specially treating keys started with `m` and `t` as the first keyspace (keyspace 0), and using the following key-value encoding:

```
// for keyspace 0
{m|t}: prefix for keyspace 0, which is already used by TiDB

// for other keyspaces
k{keyspace id}x: prefix for transactional keyspaces
k{keyspace id}r: prefix for non-transactional keyspaces
k{keyspace id}a: prefix for non-transactional keyspaces with atmoic support
```